### PR TITLE
feat: add `clone` for dataframes

### PR DESCRIPTION
**Pull request by EdAbati**
## What type of PR is this? (check all applicable)

- [ ] 💾 Refactor
- [x] ✨ Feature
- [ ] 🐛 Bug Fix
- [ ] 🔧 Optimization
- [ ] 📝 Documentation
- [ ] ✅ Test
- [ ] 🐳 Other

## Related issues 

 Mentioned in #355 

## Checklist

- [x] Code follows style guide (ruff)
- [x] Tests added 
- [x] Documented the changes

## If you have comments or can explain your changes, please do so below.

I am not very familiar with pyarrow Dataframes. What could be a clean way to implement copy/clone?
I added `NotImplementedError` as a placeholder, if we keep the error I will add a test

### CODE REVIEW

**Review by MarcoGorelli — COMMENTED** [review]
Nice, thanks @EdAbati !

I just have two comments:
- should we add LazyFrame.clone too?
- seeing as this is a new test, shall we follow the practices from https://github.com/narwhals-dev/narwhals/issues/401, and use a new `tests/frame/clone_test.py` file, and use the `constructor` fixture? i.e. the one without pyarrow

File: tests/conftest.py [hunk 1]
```diff
@@ -52,6 +52,10 @@ def polars_constructor(obj: Any) -> IntoDataFrame:
     return pl.DataFrame(obj)
 
 
+def polars_lazy_constructor(obj: Any) -> pl.LazyFrame:
```
Comment (EdAbati) [inline tests/conftest.py:55]: I think I couldn't use `IntoDataFrame` here. What would be the correct way to type this?

Should we add a `IntoLazyFrame` type too?

**Comment by EdAbati** [comment]
(BTW the test are looking very nice now with the new `pytest.fixture`s)

File: tests/conftest.py (hunk 1 above)
Comment (MarcoGorelli) [reply tests/conftest.py:55]: yup, probably! and maybe we will also need `IntoFrame` for when you don't know?

on the typing front, I'd like to see if we can make `DataFrame` generic

**Review by MarcoGorelli — APPROVED** [review]
graaaandee @EdAbati 

if there's anything you need for sklearn lmk, thanks for doing this!

**Pull request merged**